libcaf_test/caf/test/test.test.cpp

```diff
@@ -7,13 +7,33 @@
 
 using caf::test::block_type;
 
-TEST("tests can contain checks") {
+TEST("tests can contain different type of checks") {
   auto* rep = caf::test::reporter::instance;
-  for (int i = 0; i < 3; ++i)
-    check_eq(i, i);
+  /// tests all of the possible scenario for check_ne
+  check_ne(0u, 0u);
+  check_ne(0u, 1u);
   auto stats = rep->test_stats();
-  check_eq(stats.passed, 3u);
-  check_eq(stats.failed, 0u);
+  /// tests all of the possible scenario for check_eq
+  check_eq(stats.passed, 1u);
+  check_eq(stats.passed, 0u);
```

Comment:
I find these tests very confusing. Why are we checking `stats.passed` here? The `check_ne` checks above make sense and use obvious values. Also, this entire suite isn't really testing anything. At the end of it all, this test overrides the test stats without ever checking that the things that are supposed to fail actually fail.

Here's an idea:

```c++
  // Either as a lambda here or as a member function in `runnable`.
  auto should_fail = [this](auto expr, const detail::source_location& location
                = detail::source_location::current()) {
    auto before = rep->test_stats();
    expr();
    auto after = rep->test_stats();
    check_eq(before.passed, after.passed, location);
    if (check_eq(before.failed + 1, after.failed, location)) {
      rep->test_stats({before.passed + 1, before.failed});
  };
```

This gives us a clean way to single-out tests that we expect to fail and to selectively "fix" the `failed` count. What do you think? With this utility, we can write tests like this:

```suggestion
  /// Test `check_eq`.
  check_eq(1, 1);
  should_fail([this] { check_eq(1, 2); });
  should_fail([this] { check_eq(2, 1); });
```

Reply:
Yes, that makes sense as we are not confirming if the things that should fail, actually fails. But, my current understanding is that we are using these tests to verify the binary predicates themselves and I thought that adding a new class that returns a value will not be a good idea. Do you recommend adding a new class `test.cpp` which will be tested using various predicates in this file?